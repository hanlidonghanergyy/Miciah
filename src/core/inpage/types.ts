export enum JsonRpcMethod {
  // synchronous or asynchronous
  eth_chainId = "eth_chainId",
  net_version = "net_version",
  eth_accounts = "eth_accounts",
  eth_coinbase = "eth_coinbase",
  eth_uninstallFilter = "eth_uninstallFilter", // synchronous

  // asynchronous only
  eth_requestAccounts = "eth_requestAccounts",
  eth_sign = "eth_sign",
  eth_ecRecover = "eth_ecRecover",
  personal_sign = "personal_sign",
  personal_ecRecover = "personal_ecRecover",
  eth_signTransaction = "eth_signTransaction",
  eth_sendRawTransaction = "eth_sendRawTransaction",
  eth_sendTransaction = "eth_sendTransaction",
  eth_signTypedData = "eth_signTypedData",
  eth_signTypedData_v1 = "eth_signTypedData_v1",
  eth_signTypedData_v2 = "eth_signTypedData_v2",
  eth_signTypedData_v3 = "eth_signTypedData_v3",
  eth_signTypedData_v4 = "eth_signTypedData_v4",
  wallet_getPermissions = "wallet_getPermissions",
  wallet_requestPermissions = "wallet_requestPermissions",
  wallet_addEthereumChain = "wallet_addEthereumChain",
  wallet_switchEthereumChain = "wallet_switchEthereumChain",
  wallet_watchAsset = "wallet_watchAsset",

  // asynchronous pub/sub
  eth_subscribe = "eth_subscribe",
  eth_unsubscribe = "eth_unsubscribe",

  // asynchronous filter methods
  eth_newFilter = "eth_newFilter",
  eth_newBlockFilter = "eth_newBlockFilter",
  eth_newPendingTransactionFilter = "eth_newPendingTransactionFilter",
  eth_getFilterChanges = "eth_getFilterChanges",
  eth_getFilterLogs = "eth_getFilterLogs",
}

export const AUTHORIZED_METHODS = new Set<string>([
  JsonRpcMethod.eth_sign,
  JsonRpcMethod.personal_sign,
  JsonRpcMethod.eth_signTransaction,
  JsonRpcMethod.eth_signTypedData,
  JsonRpcMethod.eth_signTypedData_v1,
  JsonRpcMethod.eth_signTypedData_v2,
  JsonRpcMethod.eth_signTypedData_v3,
  JsonRpcMethod.eth_signTypedData_v4,
]);

/**
 * A String specifying the version of the JSON-RPC protocol.
 * MUST be exactly "2.0".
 */
export type JsonRpcVersion = "2.0";

/**
 * An identifier established by the Client that MUST contain a String, Number,
 * or NULL value if included. If it is not included it is assumed to be a
 * notification. The value SHOULD normally not be Null and Numbers SHOULD
 * NOT contain fractional parts.
 */
export type JsonRpcId = number | string | null;

export interface JsonRpcError {
  code: number;
  message: string;
  data?: unknown;
  stack?: string;
}

export interface JsonRpcRequest<T> {
  jsonrpc: JsonRpcVersion;
  method: string;
  id: JsonRpcId;
  params?: T;
}

export interface JsonRpcNotification<T> {
  jsonrpc: JsonRpcVersion;
  method: string;
  params?: T;
}

export type JsonRpcCallback<T> = (
  error: Error | null,
  result?: JsonRpcResponse<T>
) => void;

export type JsonRpcCallbackBatch = (
  error: Error | null,
  result?: JsonRpcResponse<unknown>[]
) => void;

interface JsonRpcResponseBase {
  jsonrpc: JsonRpcVersion;
  id: JsonRpcId;
}

export interface JsonRpcSuccess<T> extends JsonRpcResponseBase {
  result: T;
}

export interface JsonRpcFailure extends JsonRpcResponseBase {
  error: JsonRpcError;
}

export type JsonRpcResponse<T> = JsonRpcSuccess<T> | JsonRpcFailure;
