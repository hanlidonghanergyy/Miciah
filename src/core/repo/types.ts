import { AccountParams } from "core/types";

export enum RepoTable {
  Networks = "networks",
  Accounts = "accounts",
}

export interface INetwork {
  chainId: number;
  rpcUrls: string[];
  name: string;
  mainnet: boolean;
  nativeCurrency: {
    name: string;
    symbol: string; // 2-6 characters long
    decimals: 18;
  };
  blockExplorerUrls?: string[];
  iconUrls?: string[];
}

export type IAccount = AccountParams & {
  address: string;
  name: string;
  usdValues: Record<number, string>;
};
