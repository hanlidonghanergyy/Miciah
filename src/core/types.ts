export enum TakyMessageType {
  PageMessage = "PAGE_MESSAGE",
}

export interface TakyMessageBase {
  type: TakyMessageType;
}

export interface TakyPermissionMessage extends TakyMessageBase {
  type: TakyMessageType.PageMessage;
}
