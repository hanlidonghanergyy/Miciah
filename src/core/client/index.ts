import { PorterClient } from "lib/ext/porter/client";
import { assert } from "lib/system/assert";
import {
  PorterChannel,
  Request,
  Response,
  EventMessage,
  MessageType,
  WalletStatus,
  AddAccountParams,
  SeedPharse,
} from "core/types";

const porter = new PorterClient<Request, Response>(PorterChannel.Wallet);

export async function getWalletStatus() {
  const type = MessageType.GetWalletStatus;
  const res = await porter.request({ type });
  assert(res?.type === type);
  return res.status;
}

export function onWalletStatusUpdated(
  callback: (newWalletStatus: WalletStatus) => void
) {
  return porter.onMessage<EventMessage>((msg) => {
    if (msg.type === MessageType.WalletStatusUpdated) {
      callback(msg.status);
    }
  });
}

export async function setupWallet(
  password: string,
  accountParams: AddAccountParams,
  seedPhrase?: SeedPharse
) {
  const type = MessageType.SetupWallet;
  const res = await porter.request({
    type,
    password,
    accountParams,
    seedPhrase,
  });
  assert(res?.type === type);
  return res.accountAddress;
}

export async function unlockWallet(password: string) {
  const type = MessageType.UnlockWallet;
  const res = await porter.request({
    type,
    password,
  });
  assert(res?.type === type);
}

export async function lockWallet() {
  const type = MessageType.LockWallet;
  const res = await porter.request({ type });
  assert(res?.type === type);
}

export async function addAccount(params: AddAccountParams) {
  const type = MessageType.AddAccount;
  const res = await porter.request({
    type,
    params,
  });
  assert(res?.type === type);
  return res.accountAddress;
}

export async function deleteAccount(password: string, accountAddress: string) {
  const type = MessageType.DeleteAccount;
  const res = await porter.request({
    type,
    password,
    accountAddress,
  });
  assert(res?.type === type);
}

export async function isWalletHasSeedPhrase() {
  const type = MessageType.HasSeedPhrase;
  const res = await porter.request({ type });
  assert(res?.type === type);
  return res.seedPhraseExists;
}

export async function addSeedPhrase(seedPhrase: SeedPharse) {
  const type = MessageType.AddSeedPhrase;
  const res = await porter.request({
    type,
    seedPhrase,
  });
  assert(res?.type === type);
}

export async function getSeedPhrase(password: string) {
  const type = MessageType.GetSeedPhrase;
  const res = await porter.request({
    type,
    password,
  });
  assert(res?.type === type);
  return res.seedPhrase;
}

export async function getPrivateKey(password: string, accountAddress: string) {
  const type = MessageType.GetPrivateKey;
  const res = await porter.request({
    type,
    password,
    accountAddress,
  });
  assert(res?.type === type);
  return res.privateKey;
}

export async function getPublicKey(accountAddress: string) {
  const type = MessageType.GetPublicKey;
  const res = await porter.request({
    type,
    accountAddress,
  });
  assert(res?.type === type);
  return res.publicKey;
}
