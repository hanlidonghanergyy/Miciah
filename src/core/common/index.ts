export * from "./base";
export * from "./wallet";
export * from "./account";
export * from "./settings";
