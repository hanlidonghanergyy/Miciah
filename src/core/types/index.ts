export * from "./base";
export * from "./storage";
export * from "./account";
export * from "./network";
export * from "./messages";
export * from "./shared";
export * from "./rpc";
export * from "./activity";
export * from "./tokens";
export * from "./sync";
export * from "./fees";
export * from "./contacts";
export * from "./permissions";
