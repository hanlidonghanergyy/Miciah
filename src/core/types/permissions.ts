export interface Permission {
  origin: string;
  timeAt: number;
  id: string;
  chainId: number;
  accountAddresses: string[];
}
