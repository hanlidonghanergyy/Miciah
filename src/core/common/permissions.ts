import { assert } from "lib/system/assert";

import { Permission, ActivitySource } from "core/types";

export function wrapPermission(p: Permission) {
  return {
    caveats: [
      {
        type: "restrictReturnedAccounts",
        value: p.accountAddresses,
      },
    ],
    date: p.timeAt,
    id: p.id,
    invoker: p.origin,
    parentCapability: "eth_accounts",
  };
}

export function getPageOrigin(source: ActivitySource) {
  assert(source.type === "page");
  return new URL(source.url).origin;
}
