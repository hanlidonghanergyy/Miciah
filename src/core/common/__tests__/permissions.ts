import { Permission, SelfActivityKind } from "core/types";

import { wrapPermission, getPageOrigin } from "../permissions";

describe("wrapPermission", () => {
  it("should wrap permission with correct parent capability", () => {
    const permission: Permission = {
      accountAddresses: ["0x123", "0x456"],
      timeAt: 1697821898588, // Date.now() - static
      id: "123",
      origin: "test",
      chainId: 1,
    };
    const wrappedPermission = wrapPermission(permission);
    expect(wrappedPermission).toMatchSnapshot();
  });
});

describe("getPageOrigin", () => {
  it("should return page origin from activity source", () => {
    const activitySource = {
      type: "page" as const,
      url: "https://example.com/foo/bar/?x=y",
    };
    const pageOrigin = getPageOrigin(activitySource);
    expect(pageOrigin).toBe("https://example.com");
  });

  it("should throw error if activity source type is not page", () => {
    const activitySource = {
      type: "self" as const,
      kind: SelfActivityKind.Unknown,
    };
    expect(() => getPageOrigin(activitySource)).toThrowError();
  });
});
