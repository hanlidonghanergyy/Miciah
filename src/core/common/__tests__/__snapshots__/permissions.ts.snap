// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wrapPermission should wrap permission with correct parent capability 1`] = `
{
  "caveats": [
    {
      "type": "restrictReturnedAccounts",
      "value": [
        "0x123",
        "0x456",
      ],
    },
  ],
  "date": 1697821898588,
  "id": "123",
  "invoker": "test",
  "parentCapability": "eth_accounts",
}
`;
